### PR TITLE
work around SR-13945 segfault

### PR DESCRIPTION
**Pull request by marcrasi**
I filed https://bugs.swift.org/browse/SR-13945 for the underlying cause. This seems like a perfectly good workaround to get swift-apis compiling again.

### CODE REVIEW

File: Sources/TensorFlow/Layers/Recurrent.swift [hunk 1]
```diff
@@ -703,3 +687,54 @@ public typealias SimpleRNNCell = BasicRNNCell
...
+  /// encounter if you tried to implement this at the callsite using a for loop.
+  @differentiable
+  func differentiableMerging(
+    _ other: Self, mergeFunction: @differentiable (Element, Element) -> Element
+  ) -> Self {
+    zip(self, other).map { mergeFunction($0.0, $0.1) }
+  }
+
+  @derivative(of: differentiableMerging)
+  func differentiableMerging(
```
Comment (dan-zheng) [inline Sources/TensorFlow/Layers/Recurrent.swift:722]:
```suggestion
  func vjpDifferentiableMerging(
```

File: Sources/TensorFlow/Layers/Recurrent.swift [hunk 2]
```diff
@@ -703,3 +687,54 @@ public typealias SimpleRNNCell = BasicRNNCell
...
+  ) -> (value: Self, pullback: (TangentVector) -> (TangentVector, TangentVector)) {
+    let valuesWithPullbacks = zip(self, other).map {
+      valueWithPullback(at: $0.0, $0.1, in: mergeFunction)
+    }
+    let pullbacks = valuesWithPullbacks.map { $0.pullback }
+    return (
+      valuesWithPullbacks.map { $0.value },
+      { vs in
+        let resultPairs: [(Element.TangentVector, Element.TangentVector)] = zip(vs.base, pullbacks).map {
+          let (v, pb) = $0
```
Comment (dan-zheng) [inline Sources/TensorFlow/Layers/Recurrent.swift:733]:
```suggestion
        let resultPairs: [(Element.TangentVector, Element.TangentVector)] = zip(vs.base, pullbacks).map { (v, pb) in
```

File: Sources/TensorFlow/Layers/Recurrent.swift [hunk 3]
```diff
@@ -703,3 +687,54 @@ public typealias SimpleRNNCell = BasicRNNCell
...
+  func differentiableReversed() -> Self {
+    .init(self.reversed())
+  }
+
+  @derivative(of: differentiableReversed)
+  func vjpDifferentiableReversed()
+    -> (value: Self, pullback: (TangentVector) -> TangentVector)
+  {
+    return (self.differentiableReversed(), { .init(.init($0.base.reversed())) })
+  }
```
Comment (dan-zheng) [inline Sources/TensorFlow/Layers/Recurrent.swift:708]: No urgency: I wonder if this derivative could be defined as `@derivative(of: reversed)` and put in the `_Differentiation` module?

**Review by dan-zheng — APPROVED** [review]

**Review by pschuh — APPROVED** [review]

File: Sources/TensorFlow/Layers/Recurrent.swift (hunk 3 above)
Comment (marcrasi) [reply Sources/TensorFlow/Layers/Recurrent.swift:708]: That would be nice. `reversed()` returns a `ReversedCollection` so we would need to make that differentiable too.

Comment (pschuh) [reply Sources/TensorFlow/Layers/Recurrent.swift:708]: reverse() is mutating. Perhaps that could be used instead?

**Pull request merged**